Add support for SVG plots using IPython/Matplotlib

jabirali opened the pull request:
This pull request resolves issue #20:

- It fixes a bug whereby Matplotlib plots "disappear" in `comint-mime` buffers if you choose to turn on SVG output in IPython/Matplotlib.
- It also turns on SVG support in IPython by default, since this significantly improves the image quality.

For comparison with #20, the same example looks like this after this PR:
<img width="682" alt="Screenshot 2024-04-24 at 17 44 33" src="https://github.com/astoff/comint-mime/assets/271284/6599c94f-5736-4b53-8aba-497648855463">

jabirali committed:
Add support for SVG plots using IPython/Matplotlib

jabirali commented:
Additional effect of using SVG for Matplotlib plots: I just discovered that the text color appears to follow the current Emacs theme. So if you use a Matplotlib style sheet with a transparent background (or set `rcParams["figure.facecolor"] = (0, 0, 0, 0)`), then the plot text automatically changes colors when Emacs does (e.g. useful with [auto-dark](https://github.com/LionyxML/auto-dark-emacs)).
<img width="714" alt="Screenshot 2024-04-24 at 22 21 24" src="https://github.com/astoff/comint-mime/assets/271284/702a5ec5-1990-4ea2-9df1-dc78678aea89">

astoff commented:
Thanks for the PR. The first change we should certainly merge. The second probably as well, but I would like to understand the consequences of enabling SVG output. Are there any downsides? Also: `InlineBackend.figure_formats` is a list, what happens when there's more than one element in it?

jabirali commented:
Thanks for the quick response!

> Thanks for the PR. The first change we should certainly merge.

Great! That one is the most important, since the second change doesn't *necessarily* have to be in `comint-mime` itself.

> The second probably as well, but I would like to understand the consequences of enabling SVG output.

This is a valid concern, and I agree that it's good to think through the default setting carefully.

I think we basically have three options here:

- We could merge the current code, thus changing to SVG plots by default. (In that case, we can consider whether to also add a line `matplotlib.rcParams["figure.facecolor"] = (0, 0, 0, 0)` so that, so that the user gets a transparent background of the axis labels that fits their theme.)
- We could remove that line from the PR, and just mention in the `comint-mime` README that the user can add the line `get_ipython().run_line_magic("config", "InlineBackend.figure_formats = ['svg']")` to their IPython config in `~/.config/ipython/profile_default/startup/start.py` to get SVG quality plots. Since this is after all an IPython setting I think that's OK, but it might be helpful to list in the README to make it more discoverable for users that don't know about the feature. One downside of this approach is that presumably users that work over TRAMP would need to sync their IPython configuration to remote hosts to get this feature, but presumably keeping it in `comint-mime` would make it "just work" over SSH (I have not tested `comint-mime` with remote servers yet, but I assume it might work).
- We could do like VSCode, and add a user-configurable option like e.g. `comint-mime-image-format` that can be set to e.g. `"svg"`, `"png"`, etc. to "recommend" an output format to the backend (in this case, IPython). This could e.g. be implemented by adding an optional argument `format=` to `__COMINT_MIME_setup`, and then this could be transmitted from `comint-mime-setup-python`. From a user's point of view, that would perhaps be ideal?

> Are there any downsides?

I did a bit of research, and it seems that there are two potential downsides here.

Over in the VSCode issue tracker, I see that VSCode did the following:
- They implemented an option `jupyter.generateSVGPlots` that enabled SVG generation in IPython/Matplotlib, and toggled it on by default. I remember trying it a while back and that it seemed to work well for me.
- I've now learned that they have since disabled that option by default, then deprecated the feature, and by now they even removed the feature completely (see [this](https://github.com/microsoft/vscode-jupyter/issues/11168) and [this](https://github.com/microsoft/vscode-jupyter/issues/13835)). The "official" way to get SVG format back in VSCode is to customize a variable `jupyter.runStartupCommands` that passes arbitrary start-up settings to IPython, and in there you can change the image formats manually if you want.

The reason for this appears to be that some users experienced that some particular notebooks swelled up from kilobytes to gigabytes when the options were switched from PNG to SVG. I don't recall struggling with this myself, but I can easily imagine this happening if you e.g. were to plot high-density non-gridded data. Also, SVG is an XML file that will already naturally be less compact than a well-compressed PNG.

Based on the same logic as the VSCode people had, I think SVG might not be a great default then, since it would certainly be surprising to a user if Emacs spent gigabytes of memory on showing a few plots. As a user, I would however still enjoy having an option like e.g. `comint-mime-image-format` to toggle it on;  option, as the SVG looks much better on my system at least – and if the SVG size becomes a problem, then probably I should anyway change the plot code as a future PDF export is likely to have the same issue ;).

The other potential drawback is the feature I was a bit excited about above: That the font text appears to follow the current Emacs theme. While I find this a positive thing, I haven't tested it across a wide range of plots and configurations (e.g. many different Matplotlib themes, text in inset plots, different operating systems than MacOS, etc.). Depending on how consistently the SVG text is fontified, one might need to manually set the faces used in the plots in such a way that they don't result in e.g. white-on-white text.

Given all of this, my suggestion would perhaps be to not toggle it on by default then, but it could be nice to have an option like `comint-mime-image-format` as suggested above. (Or, alternatively, just amending the README to explain how the user might configure this themselves.)

What do you think?

> Also: InlineBackend.figure_formats is a list, what happens when there's more than one element in it?

Well, based on [this](https://ipython.org/ipython-doc/2/config/options/notebook.html), the documentation is a bit sparse but we can learn the following about it:

- It seems that it used to be a single string named `InlineBackend.figure_format`, but that was since replaced by a set named `InlineBackend.figure_formats`. Since its a set and not a list, we can not really specify any preference (e.g. SVG over PNG) by adding items to this config option.
- The default value of that set is `['png']`, so setting it to a single format appears to still be normal.
- The allowed values are `png, retina, jpeg, svg, pdf`. Based on this, my guess is that `retina` is the main one that perhaps makes sense to mix with the others, presumably as a different way to resolve a similar issue as the one I was having (i.e. PNG plots being blurry on Retina screens) without going to SVG.

I did a bit of testing on my machine:
- Whether we try to set the variable to `['png']`, `['png', 'svg']`, or `['svg', 'png']`, the figure remains blurry. It's only by setting it to `['svg']` that it gets crisp.
- Setting it to `['png', 'retina']` *does* change the behavior: It produces a twice as large plot, but it still looks blurry. I think however this is because Emacs on MacOS implements Retina support in raster images in a "weird" way (if I recall correctly, one needs to inform Emacs that an image is high-DPI by using either a `:data-2x` property or saving the file as `filename@2x`).
- As one might expect, `['jpg']` results in an even more blurry picture than `['png']`.
- If I write `['pdf']` the image disappears in `comint-mime`. This is probably due to the same `MIME_TYPE` issue as we had to fix when using SVG output. (But I don't think PDF is so useful for interactive use, that's more useful as an export target in my opinion, and that's something to handle in the user's code.)

Based on this, I think the conclusion is that the only values that really make sense in the context of `comint-mime` would be `['png']` for raster images or `['svg']` for scalable images. Instead of a general `comint-mime-image-format` type option, it might then make sense to have a more specific `comint-mime-image-scalable` bool as a toggle for whether we should prefer SVG output from REPLs when it is supported (with currently only Python implemented of course).

jabirali committed:
Make the IPython figure format user-configurable

jabirali commented:
Based on the considerations above, I have now updated the pull request as follows (see c7af585):

- I added a setting `comint-mime-image-scalable` (boolean) that allows the user to request SVG images instead of PNG images from REPLs. This makes it easy for e.g. me to drop a `(setq comint-mime-image-scalable t)` into my config to enable SVG plots in IPython/Matplotlib while keeping the default setting at the perhaps more robust PNG value.
- I modified `comint-mime-setup-python` on the Lisp side and `__COMINT_MIME_setup` on the Python side to transmit this user setting from Emacs to IPython, and to thus only enable the SVG plot format if the user requests it.
- I also changed the default figure background (i.e. the color of the area around the plot itself) to be transparent when SVG is enabled. That way, it automatically changes color when the user changes their theme.

I think this might perhaps be a better way to implement it than my original PR :).

astoff commented:
Thanks for the detailed information. I agree with your conclusion, so I will merge your suggestion. I just have two small requests.

1. How about renaming the variable to `comint-mime-prefer-svg`?
2. There are some tab characters, could you untabify (replace by spaces)?

I don't need any paperwork to merge this small change, but, because this package is on GNU ELPA, you would need to sign the FSF copyright assignment papers if you want to continue contributing patches.

Also so you know: I will add similar changes in [drepl](https://github.com/astoff/drepl/), a new package I'm developing.

jabirali committed:
Revised variable names and fixed indentation

astoff commented:
@jabirali FYI, for the reasons explained in #23, I removed the line of your MR that added a workaround for dark theme users; unfortunately it seems in this case the user has to fix the problem on their own.

Thankfully that's not too hard. I added the following explanation to `comint-mime-prefer-svg`, which you might want to apply:

> Using a theme with dark background can render some SVGs unreadable. If you experience this, try setting `comint-mime-image-props' to (:foreground "black" :background "white"). Alternatively, configure the SVG creator to produce images that play well with your theme.

jabirali commented:
That sound reasonable. Thanks for the heads up! :)
